### PR TITLE
Improve scheduling fairness when batching cluster state changes with equal priority

### LINKED ISSUE #20768: Scheduling fairness when batching cluster states
While performing RCA for a different tribe node behavior, Yannick identified an additional potential improvement around scheduling fairness when batching cluster states.  This can cause pending cluster state updates to accumulate.  

Summary below:

"Pending tasks in most of the diagnostic snapshots show there are pending cluster state updates from all of the upstream clusters. Batching is done on a per upstream cluster level. This means that with 22 clusters, pending tasks should in the worst case take 22 clusters * 15 seconds per cluster state update = 5,5 minutes. This seems not to be the case, pending tasks show tasks waiting for 20 minutes (and more). The issue is I think how scheduling fairness comes into play when batching cluster states.

Assume the following cluster state updates being submitted by the upstream tribe nodes to the "compound" tribe node:

```
cluster update from cluster A (A1)
cluster update from cluster B (B1)
cluster update from cluster A (A2)
cluster update from cluster B (B2)
cluster update from cluster A (A3)
cluster update from cluster C (C1)
```

Cluster state update tasks are processed using a priority queue as a work queue. If two tasks have the same priority, the first to arrive will be executed first (FIFO style). For the tribe node, all cluster state update tasks that are generated for incoming cluster state updates have the same priority, and are thus processed in FIFO order. Cluster service also has support for batching cluster state updates, explained later. The first cluster state update to be scheduled will be cluster update from cluster A. Due to batching it will process the payloads A1, A2, and A3 in one go. Even though the payloads A1, A2 and A3 are processed in one go, the tasks that had the payloads A2 and A3 are still in the queue:

```
cluster update from cluster B (B1)
cluster update from cluster A (processed)
cluster update from cluster B (B2)
cluster update from cluster A (processed)
cluster update from cluster C (C1)
```

The next cluster state update to be processed will be the one from cluster B (payloads B1 and B2), yielding the queue:

```
cluster update from cluster A (processed)
cluster update from cluster B (processed)
cluster update from cluster A (processed)
cluster update from cluster C (C1)
```

If, in the meantime, a new update from cluster A has come in, the queue looks as follows:

```
cluster update from cluster A (processed)
cluster update from cluster B (processed)
cluster update from cluster A (processed)
cluster update from cluster C (C1)
cluster update from cluster A (A4)
```

This means however that A4 gets to execute before C1 because it will be picked up by the first "cluster update from cluster A" task. If there are long sequences of these A and B before any C, it can take quite a while before C gets a chance to execute.

@bleskes @ywelsch

### PR DESCRIPTION
**Pull request by bleskes**
As the wise man @ywelsch said: currently when we batch cluster state update tasks by the same executor, we the first task un-queued from the pending task queue. That means that other tasks for the same executor are left in the queue. When those are dequeued, they will trigger another run for the same executor. This can give unfair precedence to future tasks of the same executor, even if they weren't batched in the first run. Take this queue for example (all with equal priority)

```
 T1 (executor 1)
 T2 (executor 1)
 T3 (executor 2)
 T4 (executor 2)
 T5 (executor 1)
 T6 (executor 1)
```

 If T1 & T2 are picked up first (when T5 & T6 are not yet queued), one would expect T3 & T4 to run second. However, since T2 is still in the queue, it will trigger execution of T5 & T6.

 The fix is easy - ignore processed tasks when extracting them from the queue.

Closes #20768

### CODE REVIEW

**Comment by bleskes** [comment]
@ywelsch care to take a look?

File: core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java [hunk 1]
```diff
@@ -891,7 +891,11 @@ public TimeValue ackTimeout() {
 
         @Override
         public void run() {
-            runTasksForExecutor(executor);
+            // if this task is already processed, it shouldn't trigger another executing of pending
+            // tasks for the same executor, to be fair to other executors
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java]: maybe more like this:
if this task is already processed, the executor shouldn't execute other tasks (that arrived later), to give other executors a chance to execute their tasks.

File: core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java [hunk 2]
```diff
@@ -301,6 +302,63 @@ public void onFailure(String source, Exception e) {
         assertTrue(published.get());
     }
 
+    public void testOneExecutorDontStarveAnother() throws InterruptedException {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java]: add link to Issue (which has a more detailed explanation)?

**Review by ywelsch — APPROVED** [review]
Left minor comments. Thanks for fixing this.

File: core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java [hunk 3]
```diff
@@ -301,6 +302,63 @@ public void onFailure(String source, Exception e) {
...
+            }
+        }
+
+        TaskExecutor executor1 = new TaskExecutor();
+        TaskExecutor executor2 = new TaskExecutor();
+
+        final ClusterStateTaskConfig config = ClusterStateTaskConfig.build(Priority.NORMAL);
+        final ClusterStateTaskListener noopListener = (source, e) -> { throw new AssertionError(source, e); };
+        // this blocks the cluster state queue, so we can set it up right
+        clusterService.submitStateUpdateTask("0", 0, config, executor1, noopListener);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java]: makes sense.

File: core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java [hunk 4]
```diff
@@ -301,6 +302,63 @@ public void onFailure(String source, Exception e) {
...
+        // wait to be processed
+        startedProcessing.acquire(1);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceTests.java]: good idea. This test is hairy.

File: core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java]: ++

**Comment by bleskes** [comment]
Thx @ywelsch !

**Pull request merged**